File: Problems/Leetcode/2001-3000/2701-2800/2707.py
from functools import cache
from typing import List


class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        d = set(dictionary)

        @cache
        def dfs(i: int) -> int:
            if i < 0:
                return 0

            res = dfs(i - 1) + 1
            for j in range(i + 1):
                if s[j : i + 1] in d:
                    res = min(res, dfs(j - 1))
            return res

        return dfs(len(s) - 1)

File: Problems/Leetcode/2001-3000/2701-2800/2707.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

class Solution {
public:
    int minExtraChar(string s, vector<string>& dictionary) {
        set<string> S(dictionary.begin(), dictionary.end());

        int n = s.size();
        vector<int> f(n + 1);
        for (int i = 0; i < n; ++i) {
            f[i + 1] = f[i] + 1;
            for (int j = 0; j <= i; ++j) {
                if (S.contains(s.substr(j, i - j + 1))) {
                    f[i + 1] = min(f[i + 1], f[j]);
                }
            }
        }
        return f[n];
    }
};

int main() {
    Solution sol;

    return 0;
}